apollo-federation/src/query_plan/fetch_dependency_graph.rs

```diff
@@ -2478,6 +2478,38 @@ impl FetchDependencyGraphNode {
             _ => err,
         })?;
 
+        // this function removes unnecessary pieces of the query plan requires selection set.
+        // PORT NOTE: this function was called trimSelectioNodes in the JS implementation
+        fn trim_requires_selection_set(
+            selection_set: &executable::SelectionSet,
+        ) -> Vec<executable::Selection> {
+            selection_set
+                .selections
+                .iter()
+                .filter_map(|s| match s {
+                    executable::Selection::Field(field) => Some(executable::Selection::from(
+                        executable::Field::new(field.name.clone(), field.definition.clone())
+                            .with_selections(trim_requires_selection_set(&field.selection_set)),
+                    )),
+                    executable::Selection::InlineFragment(inline_fragment) => {
+                        let new_fragment = if inline_fragment.type_condition.is_some() {
+                            executable::InlineFragment::with_type_condition(
+                                inline_fragment.type_condition.clone().unwrap(),
+                            )
+                        } else {
+                            executable::InlineFragment::without_type_condition(
+                                inline_fragment.selection_set.ty.clone(),
+                            )
+                        }
+                        .with_selections(trim_requires_selection_set(
+                            &inline_fragment.selection_set,
+                        ));
+                        Some(executable::Selection::from(new_fragment))
+                    }
```

Comment:
nit: this can be more cleanly expressed with `Option` methods:
```suggestion
                    executable::Selection::InlineFragment(inline_fragment) => {
                        let new_fragment = inline_fragment
                            .type_condition
                            .clone()
                            .map(executable::InlineFragment::with_type_condition)
                            .unwrap_or_else(|| {
                                executable::InlineFragment::without_type_condition(
                                    inline_fragment.selection_set.ty.clone(),
                                )
                            })
                            .with_selections(trim_requires_selection_set(
                                &inline_fragment.selection_set,
                            ));
                        Some(executable::Selection::from(new_fragment))
                    }
```